libgrust/libproc_macro/tokenstream.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright (C) 2023 Free Software Foundation, Inc.
+//
+// This file is part of the GNU Proc Macro Library.  This library is free
+// software; you can redistribute it and/or modify it under the
+// terms of the GNU General Public License as published by the
+// Free Software Foundation; either version 3, or (at your option)
+// any later version.
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// Under Section 7 of GPL version 3, you are granted additional
+// permissions described in the GCC Runtime Library Exception, version
+// 3.1, as published by the Free Software Foundation.
+
+// You should have received a copy of the GNU General Public License and
+// a copy of the GCC Runtime Library Exception along with this program;
+// see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+#include "tokenstream.h"
+#include "tokentree.h"
+
+#include <cstring>
+
+namespace TokenStream {
+
+TokenStream
+TokenStream::make_tokenstream (std::vector<TokenTree::TokenTree> vec)
+{
+  auto stream = make_tokenstream (vec.size ());
+  for (auto tt : vec)
+    {
+      stream.push (tt);
+    }
+  return stream;
+}
+
+TokenStream
+TokenStream::make_tokenstream (std::uint64_t capacity)
+{
+  auto *data = new TokenTree::TokenTree[capacity];
+  return {data, 0, capacity};
+}
+
+void
+TokenStream::grow (std::uint64_t delta)
+{
+  auto new_capacity = capacity + delta;
+  auto *new_data = new TokenTree::TokenTree[new_capacity];
+  std::memcpy (new_data, data, size);
+  delete[] data;
+  data = new_data;
+}
+
+void
+TokenStream::push (TokenTree::TokenTree tree)
+{
+  if (size == capacity)
+    grow (capacity);
+  data[size] = tree;
+  size++;
+}
+
+void
+TokenStream::drop (TokenStream *stream)
+{
+  for (std::uint64_t i = 0; i < stream->size; i++)
+    {
+      TokenTree::TokenTree::drop (stream->data + i);
```

Comment:
```suggestion
      TokenTree::TokenTree::drop (&stream->data[i]);
```
nit: any reason you're not doing this? I personally find it a little clearer but if you don't agree with it your version is fine as well :)

Reply:
Ok, rebased and changed.